### PR TITLE
Alerting: update authorization logic to use proper legacy roles when fine-grained access is disabled

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
When fine-grained access is disabled, the authorization logic falls back to legacy authorization, which is by default requires user to be signed in. The additional logic was implemented on request handler side (not every where). 

This PR updates the authorization method to require `Editor` role for all POST, PUT, DELETE requests by default. And Viewer role for all other request types. There are some overrides for particular paths. Some of them were added to match the legacy behavior. 

I had to return handler ReqSignedIn to the routes group because web handler `middleware.ReqEditorRole` does not check that user is not authenticated, and therefore they get wrong status code in the response.

### CODE REVIEW

File: pkg/services/ngalert/api/authorization.go [hunk 1]
```diff
@@ -136,13 +150,15 @@ func (api *API) authorize(method, path string) web.Handler {
...
 	case http.MethodGet + "/api/alertmanager/grafana/config/api/v1/alerts":
+		fallback = ac.ReqOrgAdminOrEditor // moved from the request handler as is
 		eval = ac.EvalPermission(ac.ActionAlertingNotificationsRead)
 	case http.MethodGet + "/api/alertmanager/grafana/api/v2/status":
 		eval = ac.EvalPermission(ac.ActionAlertingNotificationsRead)
 	case http.MethodPost + "/api/alertmanager/grafana/config/api/v1/alerts":
 		// additional authorization is done in the request handler
 		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingNotificationsUpdate), ac.EvalPermission(ac.ActionAlertingNotificationsCreate), ac.EvalPermission(ac.ActionAlertingNotificationsDelete))
 	case http.MethodPost + "/api/alertmanager/grafana/config/api/v1/receivers/test":
+		fallback = ac.ReqOrgAdminOrEditor // moved from the request handler as is
```
Comment (gotjosh) [inline pkg/services/ngalert/api/authorization.go:161]:
```suggestion
		fallback = ac.ReqOrgAdminOrEditor
```

Comment (gotjosh) [reply pkg/services/ngalert/api/authorization.go:161]: Not sure these comments make sense, what are you trying to convey? Will they be useful in the long run?

File: pkg/services/ngalert/api/authorization.go [hunk 2]
```diff
@@ -26,6 +26,17 @@ func (api *API) authorize(method, path string) web.Handler {
 	authorize := acmiddleware.Middleware(api.AccessControl)
 	var eval ac.Evaluator = nil
 
+	// can be overridden in path matchers below
```
Comment (gotjosh) [inline pkg/services/ngalert/api/authorization.go:29]:
```suggestion
	// Most routes follow this general authorization approach as a fallback. Exceptions are overridden directly in the below block.

File: pkg/services/ngalert/api/authorization.go [hunk 3]
```diff
@@ -136,13 +150,15 @@ func (api *API) authorize(method, path string) web.Handler {
 	case http.MethodDelete + "/api/alertmanager/grafana/config/api/v1/alerts": // reset alertmanager config to the default
 		eval = ac.EvalPermission(ac.ActionAlertingNotificationsDelete)
 	case http.MethodGet + "/api/alertmanager/grafana/config/api/v1/alerts":
+		fallback = ac.ReqOrgAdminOrEditor // moved from the request handler as is
 		eval = ac.EvalPermission(ac.ActionAlertingNotificationsRead)
```
Comment (gotjosh) [inline pkg/services/ngalert/api/authorization.go:154]: This is somewhat confusing.

My understanding is that `eval` would only contain the permission if FGAC is enabled. However, you're assigning two variables when you know ahead of time that only of them might not return something. 

While somewhat innocuous is not very obvious and can lead to harder debugging. How do you feel about something like:

```golang
type scopedAction struct {
	action string
	scopes []string
}

func evalWithFallback(actions []scopedAction, fallback string) Evaluator {
	if len(actions) == 1 {
		// evaluate one
	}

	// evaluateAny

	return fallback
}
```

With that, you could replace this to:

```golang
evalWithFallback(ac.ActionAlertingNotificationsRead, ac.ReqOrgAdminOrEditor)
```

**Review by gotjosh — COMMENTED** [review]
There's probably bigger fish to fry in `authorization.go` but please see my comment as I think it can be a quick win for routes that have a fallback.

File: pkg/services/ngalert/api/authorization.go (hunk 3 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/api/authorization.go:154]: exactly this happens several line below 
https://github.com/grafana/grafana/pull/46931/files#diff-627a793303894bbf1b53347d4c9c0316e28159fa392b13786f80553dc71e8dd4R185

This method called only when Grafana server starts and registers API routes. Once per route.

**Review by gotjosh — APPROVED** [review]

File: pkg/services/ngalert/api/authorization.go (hunk 3 above)
Comment (gotjosh) [reply pkg/services/ngalert/api/authorization.go]: Discussed offline, I misunderstood what was going on here.

**Pull request merged**